AMP List Resize: set to layout container after first render (auto-resize attribute)

cathyxz opened the pull request:
New attribute `auto-resize` for fixing #10230. The approach here is to change to layout container after first render. It works with amp-accordion, autosuggest, and under slow internet speeds where elements are loaded after first render. Not tested with proxy extension due to CORS issues, but a local autosuggest example was built for testing.

### Test Plan
- [x] Test toggle class via amp-bind use case
- [x] Test amp-accordion use case
- [x] Test change number of shown elements use case
- [x] Test load elements after render use case
- [x] Add attribute and make relevant validator changes
- [x] Test `reset-on-refresh`

### TODO
- [x] Fix unit tests
Ran unit tests with `amp-list-resize` experiment toggled on in `beforeEach`. Works for all except for the one pertaining to `attemptChangeHeight`.

aghassemi commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -387,15 +390,41 @@ export class AmpList extends AMP.BaseElement {
+      }
     }, this.container_);
+
+    if (isExperimentOn(this.getWin(), 'amp-list-resize')) {
```
possible to move this inside the existing `mutateElement` block above and remove the second `mutateElement`?

cathyxz replied:
This is because it mutates a different thing. By removing the `sizer`, we are mutating `this.element`, whereas the other mutate block is contained to `this.container_`.

aghassemi replied:
got it.

choumx replied:
You can remove `this.container_` param from the mutate call and then combine the blocks. A mutate pass encompasses children as well.

cathyxz replied:
I originally intended to preserve the limited scope because if it's size-defined, it won't need to mutate the entire element. But since the proposed experiment is to change it to be layout container... ¯\_(ツ)_/¯. 

Fixed to combine both mutate blocks.

aghassemi commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -387,15 +390,41 @@ export class AmpList extends AMP.BaseElement {
+    }
+    this.container_.classList.remove('i-amphtml-fill-content');
+    this.container_.classList.remove('i-amphtml-replaced-content');
+    const overflowElement = this.element.querySelector('[overflow]');
```
curious why this is needed. Technically after the resize `[overflow]` should become completely hidden. Whether it is left in the DOM or not should not impact anything.

cathyxz replied:
I'll investigate this. I think it was because in one of my test cases, the overflow element would not go away by itself, but I should clarify which one and why.

aghassemi commented on test/manual/amp-list-resize.amp.html:
```diff
@@ -0,0 +1,163 @@
+
+  <h2>AMP List Basic</h2>
+  <p>Test this under slow internet (throttle to slow 3G).</p>
+  <amp-list width="396" height="80" layout=responsive
```
another test with `reset-on-refresh` attribute on `amp-list` might be a good idea. (at min do a quick manual test by setting `reset-on-refresh`.

cathyxz replied:
Great idea, will do.

cathyxz replied:
Okay, the result of testing `reset-on-refresh` is that we also have to remove the `'i-amphtml-layout-size-defined'` class, which sets the placeholder to be layout fill. This way, the placeholder will take up its original size, and the `amp-list` will be sized to the the placeholder.

cathyxz replied:
Relevant changes made.

choumx commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -387,15 +390,41 @@ export class AmpList extends AMP.BaseElement {
+      }
     }, this.container_);
+
+    if (isExperimentOn(this.getWin(), 'amp-list-resize')) {
```
We only want to run this once right? Also, let's extract to local var to de-dupe with above.

cathyxz replied:
Done.

choumx commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -92,6 +94,12 @@ export class AmpList extends AMP.BaseElement {
     this.ssrTemplateHelper_ = null;
+
+    /** @private */
+    this.resizeExperimentOn_ = isExperimentOn(this.win, 'amp-list-resize');
```
Nit: Move back to `render_()` as a local var there. For testing, it's nice to be able to toggle this experiment without refreshing the page.

cathyxz replied:
Oh gotcha. Sorry I misunderstood your original comment.

cathyxz replied:
Done.

choumx commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -387,15 +396,44 @@ export class AmpList extends AMP.BaseElement {
+      }
     }, this.container_);
+
+    if (this.resizeExperimentOn_ && !this.isSetToLayoutContainer_) {
```
Consider adding a comment that this will need to be gated under an opt-in attribute.

cathyxz replied:
Done

choumx commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -387,15 +396,44 @@ export class AmpList extends AMP.BaseElement {
+      this.element.removeChild(sizer);
+    }
+    this.element.classList.remove('i-amphtml-layout-size-defined');
+    this.element.classList.remove('i-amphtml-layout-responsive');
```
Nit: `classList.remove` can take multiple args.

cathyxz replied:
TIL. xD

cathyxz replied:
Done.

choumx commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -387,15 +396,44 @@ export class AmpList extends AMP.BaseElement {
+   */
+  changeToLayoutContainer_() {
+    this.element.setAttribute('layout', 'container');
+    const sizer = this.element.querySelector('i-amphtml-sizer');
```
We should restrict this to direct children only to avoid removing a child AMP component's sizer.

cathyxz replied:
Changed to use `childElementByTag`

aghassemi commented on tools/experiments/experiments.js:
```diff
@@ -345,6 +345,11 @@ const EXPERIMENTS = [
     cleanupIssue: 'https://github.com/ampproject/amphtml/issues/16508',
   },
+  {
+    id: 'amp-list-resize',
```
I think we should add it to `doc-level optin` already since there is no API change per offline discussions.

cathyxz replied:
Ack. I think we need separate PRs now for everything that touches prod or canary config for easy rollback purposes though, right?

aghassemi replied:
oh yeah, right!

cathyxz replied:
I think we were deprecating them in favor of origin trials, which we currently don't yet have an easy process for? The intention here is to allow the issue subscribers of #10230 an easy way to opt-in so we can try and troubleshoot this before a gradual experimental launch to 100%.

cathyxz commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -387,15 +398,45 @@ export class AmpList extends AMP.BaseElement {
+        'i-amphtml-fill-content', 'i-amphtml-replaced-content');
+    const overflowElement = childElementByAttr(this.element, 'overflow');
+    if (overflowElement) {
+      toggle(overflowElement, false);
```
Unfortunately, this doesn't work for the same reason I mentioned to Ali above. The overflow element is hidden via `visibility: hidden`, but it takes up a blank space, which looks odd since we are now in layout container. Technically after we switch to layout container, we shouldn't need the `overflow` element anymore...

choumx replied:
Gotcha, thanks. Can you add a comment explaining this? Also use `getOverflowElement()`.

cathyxz replied:
Done

cathyxz commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -387,15 +398,45 @@ export class AmpList extends AMP.BaseElement {
+    if (resizeExperimentOn && !this.isSetToLayoutContainer_) {
+      this.mutateElement(() => {
+        this.changeToLayoutContainer_();
+      });
```
Done.

choumx commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -404,39 +409,30 @@ export class AmpList extends AMP.BaseElement {
+      this.element.removeChild(sizer);
+    }
+    this.element.classList.remove(
+        'i-amphtml-layout-size-defined', 'i-amphtml-layout-responsive');
```
Removing `i-amphtml-layout-responsive` is right if `layout` was `responsive`, but what about the other [supported layouts](https://www.ampproject.org/docs/reference/components/amp-list)? Check out `getLayoutClass()` in layout.js. 

Looks like `nodisplay` is also supported but we probably shouldn't change that (make it visible suddenly). 

And while we're at it, should we be adding `i-amphtml-layout-container`? Basically undoing and reapplying `layout.js#applyStaticLayout()`. There are other things that need to be undone too for other layouts e.g. removing `width` and `height` styles.

Sorry to keep piling stuff on. 💩

cathyxz replied:
It seems like for layout fill and no-display, we shouldn't really do anything here.

cathyxz replied:
Another TODO: I realized that this implementation does not play nice with reset-on-refresh because the loader is also `visibility: hidden` instead of `display: none`, which will squash out the contents of the list. We may need to toggle it to false, and possibly add additional treatment to bring it back.

cathyxz replied:
On further testing, the issue with `reset-on-refresh` only applies to `flex-item` layout. It's possible that we could bail here and not handle it (i.e. revert to old approach). The main issue here is the loader taking up the full space of the list contents, and pushing the list contents below the list (so layout is completely weird). 

Options for solving this problem: 
1. Do nothing, revert to old approach. 
2. Toggle loader to false. (Bring it back on refresh?)

cathyxz replied:
Actually it's not possible to override this because it has `display: block !important`, so maybe what we want to do is rethink the approach to reusable loaders. This seems to be a smaller edge case and possibly worth separating into a different PR since it may also touch loader logic. I've reverted `flex-item` layout to use the original approach as a stop gap for this PR.

aghassemi replied:
sgtm, `flex-item` is not commonly used and we can deal with it later.

cathyxz commented:
*blocking because I can't request changes on my own PR, but don't want this to get accidentally merged.

cathyxz commented:
@choumx @aghassemi PTAL?

aghassemi commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -386,16 +390,76 @@ export class AmpList extends AMP.BaseElement {
+  changeToLayoutContainer_() {
+    const layout = this.element.getAttribute('layout');
+    if (layout == Layout.RESPONSIVE) {
+      const sizer = childElementByTag(this.element, 'i-amphtml-sizer');
```
nit: this seems fine to do regardless of layout, similar to how you remove `'i-amphtml-layout-size-defined'` class regardless of layout after the `if else` block

cathyxz replied:
Done.

aghassemi commented on tools/experiments/experiments.js:
```diff
@@ -338,6 +338,16 @@ const EXPERIMENTS = [
     cleanupIssue: 'https://github.com/ampproject/amphtml/issues/16508',
   },
+  {
+    id: 'amp-carousel-new-arrows',
```
`amp-carousel-new-arrows` looks like a bad merge issue.

cathyxz replied:
Yup it was, reverted.

aghassemi commented:
let's document the new attribute in `md` after this makes it to prod.

cathyxz commented:
Added a comment for that in https://github.com/ampproject/amphtml/issues/10230#issuecomment-414860131.

cathyxz commented:
@choumx PTAL? ^_^

cathyxz commented:
Ping @choumx in case you wanted to take another look. 

Main change was addressing all of your comments by individually considering all separate layouts, and changing this to an attribute-gated behavior instead of an experiment (attribute is `auto-resize`). 

Documentation will follow when change hits production. 

Hoping to merge this before next canary. =)

choumx commented:
Will take a look today, sorry for the delay.

choumx commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -386,16 +389,71 @@ export class AmpList extends AMP.BaseElement {
+    this.element.classList.remove('i-amphtml-layout-size-defined');
+    this.container_.classList.remove(
+        'i-amphtml-fill-content',
+        'i-amphtml-replaced-content'
```
I'm not sure we want to remove these (undoing `applyFillContent`). Do we need to?

cathyxz replied:
Yes we do if we are setting the `<amp-list>` to layout container. `i-amphtml-fill-content` basically does this: 
```
.i-amphtml-fill-content {
    display: block;
    height: 0;
    max-height: 100%;
    max-width: 100%;
    min-height: 100%;
    min-width: 100%;
    width: 0;
    margin: auto;
}
```

Since `<amp-list>` is now layout container, it results in the `<amp-list>` having a height/width of 0, and the contents of `<amp-list>` become overflow elements.

honeybadgerdontcare commented on extensions/amp-list/validator-amp-list.protoascii:
```diff
@@ -98,6 +98,10 @@ tags: {  # <amp-list>
   disallowed_ancestor: "AMP-LIST"
   disallowed_ancestor: "AMP-STATE"
   disallowed_ancestor: "TEMPLATE"
+  attrs {
```
Validation change looks good once  line 101 is `attrs: {`

cathyxz replied:
Bad mistake. Fixed.

cathyxz commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -386,16 +389,71 @@ export class AmpList extends AMP.BaseElement {
+    const sizer = childElementByTag(this.element, 'i-amphtml-sizer');
+    if (sizer) {
+      this.element.removeChild(sizer);
+    }
```
Then can we directly reference `this.element.sizerElement` here instead of using `childElementByTag`?

cathyxz replied:
Actually typecheck doesn't let us do this. 
```
extensions/amp-list/0.1/amp-list.js:469: ERROR - Access to package-private property sizerElement of Element not allowed here.
      this.element.sizerElement = undefined;
```

cathyxz replied:
Does it make more sense to toggle this to `display: none` instead of removing it? @choumx

choumx replied:
I think we should change its visibility or you can try using `this.element.changeSize()` with no params instead.

cathyxz replied:
Corrected and tested substituting this with `this.element.changeSize()`--works. Thank you!

cathyxz commented on extensions/amp-list/0.1/test/validator-amp-list.out:
```diff
@@ -84,33 +84,41 @@ amp-list/0.1/test/validator-amp-list.html:43:2 The attribute '[state]' in tag 'a
+|    <!-- Valid amp-list with auto-resize attribute -->
+|    <amp-list width=10 height=10
+>>   ^~~~~~~~~
+amp-list/0.1/test/validator-amp-list.html:85:2 The attribute 'auto-resize' may not appear in tag 'amp-list'. (see https://www.ampproject.org/docs/reference/components/amp-list) [DISALLOWED_HTML]
```
Fixed and now this is gone.

cathyxz commented on build-system/countries.json:
```diff
@@ -0,0 +1,979 @@
+{
```
Done

cathyxz commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -386,16 +389,71 @@ export class AmpList extends AMP.BaseElement {
+    this.container_.classList.remove(
+        'i-amphtml-fill-content',
+        'i-amphtml-replaced-content'
+    );
```
Done.

cathyxz commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -386,16 +389,71 @@ export class AmpList extends AMP.BaseElement {
+          if (autoResize) {
+            // This needs to be gated behind the experiment 'amp-list-resize'
+            const layout = this.element.getAttribute('layout');
+            if (layout == Layout.NODISPLAY || layout == Layout.FILL) {
```
Done.

cathyxz commented:
@choumx PTAL?

cathyxz commented on extensions/amp-list/0.1/amp-list.js:
```diff
@@ -411,16 +414,80 @@ export class AmpList extends AMP.BaseElement {
+        this.element.classList.remove('i-amphtml-layout-intrinsic');
+        break;
+    }
+    this.element./*OK*/changeSize();
```
It cleanly removes the `sizer` element, but doesn't actually set height or width via inline CSS unless height and width are passed in as parameters. Added a comment to that effect.
